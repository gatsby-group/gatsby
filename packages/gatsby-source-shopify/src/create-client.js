import { GraphQLClient } from "graphql-request"
/**
 * Create a Shopify Storefront GraphQL client for the provided name and token.
 */
export const createClient = (shopName, accessToken) => {
  let url;
  if (shopName.indexOf("." > -1)) {
    url = `https://${shopName}/api/graphql`;
  }
  else {
    url = `https://${shopName}.myshopify.com/api/graphql`;
  }  
  return new GraphQLClient(url, {
    headers: {
      "X-Shopify-Storefront-Access-Token": accessToken,
    },
  })
}
