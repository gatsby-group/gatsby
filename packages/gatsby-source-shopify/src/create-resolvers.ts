import { CreateResolversArgs } from "gatsby"
import { getGatsbyImageFieldConfig } from "gatsby-plugin-image/graphql-utils"

import { makeResolveGatsbyImageData } from "./resolve-gatsby-image-data"

export function createResolvers(
  { createResolvers, cache }: CreateResolversArgs,
  { downloadImages, typePrefix = `` }: IShopifyPluginOptions
): void {
  if (!downloadImages) {
    createResolvers({
      [`${typePrefix}ShopifyImage`]: {
        gatsbyImageData: {
          ...getGatsbyImageFieldConfig(makeResolveGatsbyImageData(cache)),
          type: `JSON`, // Because gatsbyImageData will be undefined for GIFs it must be optional
        },
      },
    })
  }
}
