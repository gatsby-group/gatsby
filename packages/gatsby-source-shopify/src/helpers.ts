import { NodePluginArgs, SourceNodesArgs } from "gatsby"
import { createRemoteFileNode } from "gatsby-source-filesystem"

import { shopifyTypes } from "./shopify-types"

const pattern = /^gid:\/\/shopify\/(\w+)\/(.+)$/

const { CI, GATSBY_CLOUD, GATSBY_IS_PR_BUILD, NETLIFY, CONTEXT } = process.env

export function createNodeId(
  shopifyId: string,
  gatsbyApi: NodePluginArgs,
  { typePrefix = `` }: IShopifyPluginOptions
): string {
  return gatsbyApi.createNodeId(`${typePrefix}${shopifyId}`)
}

export function getPluginStatus(gatsbyApi: NodePluginArgs): any {
  return (
    gatsbyApi.store.getState().status.plugins?.[`gatsby-source-shopify`] || {}
  )
}

export function getLastBuildTime(
  gatsbyApi: NodePluginArgs,
  pluginOptions: IShopifyPluginOptions
): Date | undefined {
  const { typePrefix = `` } = pluginOptions
  const status = getPluginStatus(gatsbyApi)

  return (
    status[`lastBuildTime${typePrefix}`] &&
    new Date(status[`lastBuildTime${typePrefix}`])
  )
}

export function setCurrentBuildTime(
  gatsbyApi: NodePluginArgs,
  pluginOptions: IShopifyPluginOptions
): void {
  const { typePrefix = `` } = pluginOptions
  const status = getPluginStatus(gatsbyApi)

  gatsbyApi.actions.setPluginStatus({
    ...status,
    [`currentBuildTime${typePrefix}`]: Date.now(),
  })
}

export function setLastBuildTime(
  gatsbyApi: NodePluginArgs,
  pluginOptions: IShopifyPluginOptions
): void {
  const { typePrefix = `` } = pluginOptions
  const status = getPluginStatus(gatsbyApi)

  gatsbyApi.actions.setPluginStatus({
    ...status,
    [`currentBuildTime${typePrefix}`]: undefined,
    [`lastBuildTime${typePrefix}`]: status[`currentBuildTime${typePrefix}`],
  })
}

export function isPriorityBuild(pluginOptions: IShopifyPluginOptions): boolean {
  const isGatsbyCloudPriorityBuild =
    CI === `true` && GATSBY_CLOUD === `true` && GATSBY_IS_PR_BUILD !== `true`

  const isNetlifyPriorityBuild =
    CI === `true` && NETLIFY === `true` && CONTEXT === `production`

  return pluginOptions.prioritize !== undefined
    ? pluginOptions.prioritize
    : isGatsbyCloudPriorityBuild || isNetlifyPriorityBuild
}

export function isShopifyId(shopifyId: string): boolean {
  return pattern.test(shopifyId)
}

export function parseShopifyId(shopifyId: string): Array<string> {
  return shopifyId.match(pattern) || []
}

export function decorateBulkObject(input: IBulkResult): IDecoratedResult {
  const obj: { id?: string; shopifyId?: string; [key: string]: any } = {
    ...input,
  }

  if (typeof obj === `object`) {
    if (obj.id) {
      obj.shopifyId = obj.id
      delete obj.id
    }

    for (const key of Object.keys(obj)) {
      if (typeof obj[key] === `object`) obj[key] = decorateBulkObject(obj[key])
    }
  }

  return obj as IShopifyNode
}

export async function processShopifyImages(
  {
    actions: { createNode },
    createNodeId,
    cache,
    store,
    reporter,
  }: SourceNodesArgs,
  node: IShopifyNode
): Promise<void> {
  const type = parseShopifyId(node.shopifyId)[1]
  const imageFields = shopifyTypes[type].imageFields

  if (imageFields) {
    for (const fieldPath of imageFields) {
      const image = fieldPath
        .split(`.`)
        .reduce((acc, value) => acc[value], node)

      if (image && parseImageExtension(image.originalSrc) !== `gif`) {
        const fileNode = await createRemoteFileNode({
          url: image.originalSrc,
          cache,
          createNode,
          createNodeId,
          parentNodeId: node.id,
          store,
          reporter,
        })

        image.localFile___NODE = fileNode.id
      }
    }
  }
}

export function parseImageExtension(url: string): string {
  const basename = url.split(`?`)[0]
  const dot = basename.lastIndexOf(`.`)

  if (dot !== -1) {
    return basename.slice(dot + 1)
  } else {
    throw new Error(
      `Could not parse file extension from Shopify image URL: ${url}`
    )
  }
}
