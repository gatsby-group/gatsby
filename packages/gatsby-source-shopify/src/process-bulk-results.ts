import { SourceNodesArgs } from "gatsby"

import { shopifyTypes } from "./shopify-types"
import {
  isShopifyId,
  createNodeId,
  parseShopifyId,
  decorateBulkObject,
  processShopifyImages,
} from "./helpers"

interface IBulkResultChildren {
  [key: string]: Array<string | IBulkResult>
}

export async function processBulkResults(
  gatsbyApi: SourceNodesArgs,
  pluginOptions: IShopifyPluginOptions,
  results: BulkResults
): Promise<number> {
  let nodeCount = 0
  const promises: Array<Promise<void>> = []
  const children: IBulkResultChildren = {}

  for (let i = results.length - 1; i >= 0; i -= 1) {
    const result = decorateBulkObject(results[i])
    const resultIsNode =
      result.shopifyId &&
      JSON.stringify(Object.keys(result)) !== `["shopifyId","__parentId"]`

    if (resultIsNode) {
      const type = parseShopifyId(result.shopifyId)[1]
      const imageFields = shopifyTypes[type].imageFields
      const referenceFields = shopifyTypes[type].referenceFields

      // Only increment for parent nodes
      if (!result.__parentId) nodeCount++

      // Attach reference fields
      if (referenceFields) {
        for (const referenceField of referenceFields) {
          result[referenceField] = []
        }
      }

      // Attach children references / objects
      if (children[result.shopifyId]) {
        for (const child of children[result.shopifyId]) {
          const childType =
            typeof child === `string`
              ? parseShopifyId(child)[1]
              : child.__typename
          const field = shopifyTypes[childType].key

          if (field) {
            result[field] = [
              typeof child === `string`
                ? createNodeId(child, gatsbyApi, pluginOptions)
                : child,
              ...(result[field] || []),
            ]
          } else {
            throw new Error(
              `There is no shopifyType for child type: ${childType}`
            )
          }
        }
      }

      if (!isShopifyId(result.shopifyId)) {
        throw new Error(
          `Expected an ID in the format gid://shopify/<typename>/<id>`
        )
      }

      const node = {
        ...result,
        id: createNodeId(result.shopifyId, gatsbyApi, pluginOptions),
        internal: {
          type: `${pluginOptions.typePrefix || ``}Shopify${type}`,
          contentDigest: gatsbyApi.createContentDigest(result),
        },
      } as IShopifyNode

      if (pluginOptions.downloadImages && imageFields) {
        promises.push(
          processShopifyImages(gatsbyApi, node).then(() =>
            gatsbyApi.actions.createNode(node)
          )
        )
      } else {
        gatsbyApi.actions.createNode(node)
      }
    }

    // If the object is a child store the reference
    if (result.__parentId) {
      children[result.__parentId] = [
        ...(children[result.__parentId] || []),
        result.shopifyId || result,
      ]
    }
  }

  await Promise.all(promises)
  return nodeCount
}
