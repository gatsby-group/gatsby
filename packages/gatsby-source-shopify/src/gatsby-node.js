import { pipe } from "lodash/fp"
import chalk from "chalk"
import { forEach } from "p-iteration"
import { printGraphQLError, queryAll, queryOnce } from "./lib"
import { createClient } from "./create-client"

import {
  ArticleNode,
  BlogNode,
  CollectionNode,
  CommentNode,
  ProductNode,
  ProductOptionNode,
  ProductVariantNode,
  ProductMetafieldNode,
  ProductVariantMetafieldNode,
  ShopPolicyNode,
  ShopDetailsNode,
  PageNode,
} from "./nodes"
import {
  SHOP,
  CONTENT,
  NODE_TO_ENDPOINT_MAPPING,
  ARTICLE,
  BLOG,
  COLLECTION,
  PRODUCT,
  SHOP_POLICY,
  SHOP_DETAILS,
  PAGE,
} from "./constants"
import {
  ARTICLES_QUERY,
  BLOGS_QUERY,
  COLLECTIONS_QUERY,
  PRODUCTS_QUERY,
  SHOP_POLICIES_QUERY,
  SHOP_DETAILS_QUERY,
  PAGES_QUERY,
} from "./queries"

export const sourceNodes = async (
  {
    actions: { createNode, touchNode },
    createNodeId,
    store,
    cache,
    getCache,
    reporter,
  },
  {
    shopName,
    accessToken,
    apiVersion = `2020-07`,
    verbose = true,
    paginationSize = 250,
    languages = [`en`],
    includeCollections = [SHOP, CONTENT],
    shopifyQueries = {},
  }
) => {
  const client = locale =>
    createClient(shopName, accessToken, apiVersion, locale)

  const defaultQueries = {
    articles: ARTICLES_QUERY,
    blogs: BLOGS_QUERY,
    collections: COLLECTIONS_QUERY,
    products: PRODUCTS_QUERY,
    shopPolicies: SHOP_POLICIES_QUERY,
    shopDetails: SHOP_DETAILS_QUERY,
    pages: PAGES_QUERY,
  }

  const queries = { ...defaultQueries, ...shopifyQueries }

  // Convenience function to namespace console messages.
  const formatMsg = msg =>
    chalk`\n{blue gatsby-source-shopify/${shopName}} ${msg}`

  try {
    console.log(formatMsg(`starting to fetch data from Shopify`))

    // Arguments used for file node creation.
    const imageArgs = {
      createNode,
      createNodeId,
      touchNode,
      store,
      cache,
      getCache,
      reporter,
    }

    // Arguments used for node creation.
    const args = {
      client,
      createNode,
      createNodeId,
      formatMsg,
      verbose,
      imageArgs,
      paginationSize,
      queries,
      languages,
    }

    // Message printed when fetching is complete.
    const msg = formatMsg(`finished fetching data from Shopify`)

    let promises = []
    if (includeCollections.includes(SHOP)) {
      promises = promises.concat([
        createNodes(COLLECTION, queries.collections, CollectionNode, args),
        createNodes(
          PRODUCT,
          queries.products,
          ProductNode,
          args,
          async (product, productNode, locale) => {
            if (product.variants)
              await forEach(product.variants.edges, async edge => {
                const v = edge.node
                if (v.metafields)
                  await forEach(v.metafields.edges, async edge =>
                    createNode(
                      await ProductVariantMetafieldNode(
                        imageArgs,
                        locale
                      )(edge.node)
                    )
                  )

                return createNode(
                  await ProductVariantNode(
                    imageArgs,
                    productNode,
                    locale
                  )(edge.node)
                )
              })

            if (product.metafields)
              await forEach(product.metafields.edges, async edge =>
                createNode(
                  await ProductMetafieldNode(imageArgs, locale)(edge.node)
                )
              )

            if (product.options)
              await forEach(product.options, async option =>
                createNode(await ProductOptionNode(imageArgs, locale)(option))
              )
          }
        ),
        createShopPolicies(args),
        createShopDetails(args),
      ])
    }
    if (includeCollections.includes(CONTENT)) {
      promises = promises.concat([
        createNodes(BLOG, queries.blogs, BlogNode, args),
        createNodes(
          ARTICLE,
          queries.articles,
          ArticleNode,
          args,
          async (article, _, locale) => {
            if (article.comments)
              await forEach(article.comments.edges, async edge =>
                createNode(await CommentNode(imageArgs, locale)(edge.node))
              )
          }
        ),
        createPageNodes(PAGE, queries.pages, PageNode, args),
      ])
    }

    console.time(msg)
    await Promise.all(promises)
    console.timeEnd(msg)
  } catch (e) {
    console.error(chalk`\n{red error} an error occurred while sourcing data`)

    // If not a GraphQL request error, let Gatsby print the error.
    if (!e.hasOwnProperty(`request`)) throw e

    printGraphQLError(e)
  }
}

/**
 * Fetch and create nodes for the provided endpoint, query, and node factory.
 */
// createNodes(COLLECTION, queries.collections, CollectionNode, args),
const createNodes = async (
  endpoint,
  query,
  nodeFactory,
  {
    client,
    createNode,
    formatMsg,
    verbose,
    imageArgs,
    paginationSize,
    languages,
  },
  f = async () => {}
) => {
  // Message printed when fetching is complete.
  const msg = formatMsg(`fetched and processed ${endpoint} nodes`)

  if (verbose) console.time(msg)
  await forEach(
    languages,
    async locale =>
      await forEach(
        await queryAll(
          client(locale),
          [NODE_TO_ENDPOINT_MAPPING[endpoint]],
          query,
          paginationSize
        ),
        async entity => {
          const node = await nodeFactory(imageArgs, locale)(entity)
          createNode(node)
          await f(entity, node, locale)
        }
      )
  )
  if (verbose) console.timeEnd(msg)
}

/**
 * Fetch and create nodes for shop policies.
 */
const createShopDetails = async ({
  client,
  createNode,
  formatMsg,
  verbose,
  queries,
  languages,
}) => {
  // // Message printed when fetching is complete.
  const msg = formatMsg(`fetched and processed ${SHOP_DETAILS} nodes`)

  if (verbose) console.time(msg)
  await forEach(languages, async locale => {
    const { shop } = await queryOnce(client(locale), queries.shopDetails)
    const node = await ShopDetailsNode(locale)(shop)
    createNode(node)
  })
  if (verbose) console.timeEnd(msg)
}

/**
 * Fetch and create nodes for shop policies.
 */
const createShopPolicies = async ({
  client,
  createNode,
  formatMsg,
  verbose,
  queries,
  languages,
}) => {
  // Message printed when fetching is complete.
  const msg = formatMsg(`fetched and processed ${SHOP_POLICY} nodes`)

  if (verbose) console.time(msg)
  await forEach(languages, async locale => {
    const { shop: policies } = await queryOnce(
      client(locale),
      queries.shopPolicies
    )

    Object.entries(policies)
      .filter(([_, policy]) => Boolean(policy))
      .forEach(
        pipe(async ([type, policy]) => {
          const node = await ShopPolicyNode(locale, type)(policy)
          createNode(node)
        })
      )
  })
  if (verbose) console.timeEnd(msg)
}

const createPageNodes = async (
  endpoint,
  query,
  nodeFactory,
  { client, createNode, formatMsg, verbose, paginationSize, languages },
  f = async () => {}
) => {
  // Message printed when fetching is complete.
  const msg = formatMsg(`fetched and processed ${endpoint} nodes`)

  if (verbose) console.time(msg)
  await forEach(languages, async locale => {
    await forEach(
      await queryAll(
        client(locale),
        [NODE_TO_ENDPOINT_MAPPING[endpoint]],
        query,
        paginationSize
      ),
      async entity => {
        const node = await nodeFactory(locale)(entity)
        createNode(node)
        await f(entity)
      }
    )
  })
  if (verbose) console.timeEnd(msg)
}
