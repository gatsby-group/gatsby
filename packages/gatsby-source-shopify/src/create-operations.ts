import { SourceNodesArgs } from "gatsby"
import { shiftLeft } from "shift-left"

import { createGraphqlClient } from "./clients"
import { getLastBuildTime } from "./helpers"
import { OperationError } from "./errors"

import { ProductsQuery } from "./query-builders/products-query"
import { ProductVariantsQuery } from "./query-builders/product-variants-query"
import { CollectionsQuery } from "./query-builders/collections-query"
import { OrdersQuery } from "./query-builders/orders-query"
import { LocationsQuery } from "./query-builders/locations-query"

import {
  OPERATION_STATUS_QUERY,
  OPERATION_BY_ID,
  CANCEL_OPERATION,
} from "./static-queries"

interface Operations {
  productsOperation: ShopifyBulkOperation
  productVariantsOperation: ShopifyBulkOperation
  ordersOperation: ShopifyBulkOperation
  collectionsOperation: ShopifyBulkOperation
  locationsOperation: ShopifyBulkOperation
  cancelOperationInProgress: () => Promise<void>
  cancelOperation: (id: string) => Promise<BulkOperationCancelResponse>
  finishLastOperation: () => Promise<void>
  completedOperation: (
    operationId: string,
    interval?: number
  ) => Promise<{ node: BulkOperationNode }>
}

const finishedStatuses = [`COMPLETED`, `FAILED`, `CANCELED`, `EXPIRED`]
const failedStatuses = [`FAILED`, `CANCELED`]

export function createOperations(
  gatsbyApi: SourceNodesArgs,
  pluginOptions: ShopifyPluginOptions
): Operations {
  const graphqlClient = createGraphqlClient(pluginOptions)
  const lastBuildTime = getLastBuildTime(gatsbyApi, pluginOptions)
  const operationNamePrefix = lastBuildTime ? 'INCREMENTAL_' : ''

  function createOperation(
    operationQuery: string,
    name: string
  ): ShopifyBulkOperation {
    return {
      execute: (): Promise<BulkOperationRunQueryResponse> =>
        graphqlClient.request<BulkOperationRunQueryResponse>(operationQuery),
      name: `${operationNamePrefix}${name}`,
    }
  }

  function currentOperation(): Promise<CurrentBulkOperationResponse> {
    return graphqlClient.request(OPERATION_STATUS_QUERY)
  }

  async function finishLastOperation(): Promise<void> {
    let { currentBulkOperation } = await currentOperation()
    if (currentBulkOperation && currentBulkOperation.id) {
      const timer = gatsbyApi.reporter.activityTimer(
        `Waiting for operation ${currentBulkOperation.id} : ${currentBulkOperation.status}`
      )
      timer.start()

      while (!finishedStatuses.includes(currentBulkOperation.status)) {
        await new Promise(resolve => setTimeout(resolve, 1000))
        currentBulkOperation = (await currentOperation()).currentBulkOperation
        timer.setStatus(
          `Polling operation ${currentBulkOperation.id} : ${currentBulkOperation.status}`
        )
      }

      timer.end()
    }
  }

  async function cancelOperation(
    id: string
  ): Promise<BulkOperationCancelResponse> {
    return graphqlClient.request<BulkOperationCancelResponse>(CANCEL_OPERATION, {
      id,
    })
  }

  async function cancelOperationInProgress(): Promise<void> {
    let { currentBulkOperation: bulkOperation } = await currentOperation()
    if (!bulkOperation) {
      return
    }

    const cancelTimer = gatsbyApi.reporter.activityTimer(
      `Canceling previous operation: ${bulkOperation.id}`
    )

    cancelTimer.start()

    if (bulkOperation.status === `RUNNING`) {
      cancelTimer.setStatus(
        `Canceling a currently running operation: ${bulkOperation.id}, this could take a few moments`
      )

      const { bulkOperationCancel } = await cancelOperation(bulkOperation.id)

      bulkOperation = bulkOperationCancel.bulkOperation

      while (bulkOperation.status !== `CANCELED`) {
        await new Promise(resolve => setTimeout(resolve, 100))
        const currentOp = await currentOperation()
        bulkOperation = currentOp.currentBulkOperation
        cancelTimer.setStatus(
          `Waiting for operation to cancel: ${bulkOperation.id}, ${bulkOperation.status}`
        )
      }
    } else {
      /**
       * Just because it's not running doesn't mean it's done. For
       * example, it could be CANCELING. We still have to wait for it
       * to be officially finished before we start a new one.
       */
      while (!finishedStatuses.includes(bulkOperation.status)) {
        await new Promise(resolve => setTimeout(resolve, 100))
        bulkOperation = (await currentOperation()).currentBulkOperation
        cancelTimer.setStatus(
          `Waiting for operation to cancel: ${bulkOperation.id}, ${bulkOperation.status}`
        )
      }
    }

    cancelTimer.end()
  }

  /* Maybe the interval should be adjustable, because users
   * with larger data sets could easily wait longer. We could
   * perhaps detect that the interval being used is too small
   * based on returned object counts and iteration counts, and
   * surface feedback to the user suggesting that they increase
   * the interval.
   */
  async function completedOperation(
    operationId: string,
    interval = 1000
  ): Promise<{ node: BulkOperationNode }> {
    let operation = await graphqlClient.request<{
      node: BulkOperationNode
    }>(OPERATION_BY_ID, {
      id: operationId,
    })

    const completedTimer = gatsbyApi.reporter.activityTimer(
      `Waiting for bulk operation to complete`
    )

    completedTimer.start()

    let waitForOperation = true

    while (waitForOperation) {
      if (failedStatuses.includes(operation.node.status)) {
        completedTimer.end()
        waitForOperation = false
        throw new OperationError(operation.node)
      }

      if (operation.node.status === `COMPLETED`) {
        completedTimer.end()
        waitForOperation = false
        return operation
      }

      await new Promise(resolve => setTimeout(resolve, interval))

      operation = await graphqlClient.request<{
        node: BulkOperationNode
      }>(OPERATION_BY_ID, {
        id: operationId,
      })

      completedTimer.setStatus(shiftLeft`
        Polling bulk operation: ${operation.node.id}
        Status: ${operation.node.status}
        Object count: ${operation.node.objectCount}
      `)
    }

    throw new Error(`It should never reach this error`)
  }

  return {
    productsOperation: createOperation(
      new ProductsQuery(pluginOptions).query(),
      `PRODUCTS`
    ),

    productVariantsOperation: createOperation(
      new ProductVariantsQuery(pluginOptions).query(),
      `PRODUCT_VARIANTS`,
    ),

    ordersOperation: createOperation(
      new OrdersQuery(pluginOptions).query(),
      `ORDERS`
    ),

    collectionsOperation: createOperation(
      new CollectionsQuery(pluginOptions).query(),
      `COLLECTIONS`
    ),

    locationsOperation: createOperation(
      new LocationsQuery(pluginOptions).query(),
      `LOCATIONS`
    ),

    cancelOperationInProgress,
    cancelOperation,
    finishLastOperation,
    completedOperation,
  }
}
