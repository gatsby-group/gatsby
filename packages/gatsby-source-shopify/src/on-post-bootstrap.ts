import { NodePluginArgs } from "gatsby"

import { setLastBuildTime } from "./helpers"

export function onPostBootstrap(
  gatsbyApi: NodePluginArgs,
  pluginOptions: IShopifyPluginOptions
): void {
  setLastBuildTime(gatsbyApi, pluginOptions)
}
