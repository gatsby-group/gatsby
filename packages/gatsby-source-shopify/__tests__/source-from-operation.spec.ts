// eslint-disable-next-line @typescript-eslint/triple-slash-reference
/// <reference path="../types/interface.d.ts" />

import fs from "fs"
import path from "path"
import os from "os"
import { graphql, rest } from "msw"
import { setupServer } from "msw/node"

import { makeSourceFromOperation } from "../src/source-from-operation"
import * as processBulkResultsModule from "../src/process-bulk-results"
import { createOperations } from "../src/create-operations"

import { mockGatsbyApi, mockPluginOptions, mockBulkResults } from "./fixtures"

const BULK_DATA_URL = `http://bulk-query-data.co`

const server = setupServer()

// @ts-ignore because these types will never match
global.setTimeout = (fn: Promise<void>): Promise<void> => fn()

const nodeTypes = [`products`, `variants`, `collections`, `orders`, `locations`]

beforeAll(() => {
  server.listen()
})

afterEach(() => {
  server.resetHandlers()
})

afterAll(() => {
  server.close()
})

function resolve(data) {
  return (req, res, ctx): void => res(ctx.data(data))
}

function resolveOnce(data) {
  return (req, res, ctx): void => res.once(ctx.data(data))
}

function generateTestName(prioritize, type): string {
  const modifiers = [prioritize ? `priority` : `non-priority`]
  return `Returns the correct ${type} when running a ${modifiers[0]} build`
}

function countJsonLRows(s: string): number {
  const matches = s.match(/}$/gm) || []
  return matches.length
}

describe(`makeSourceFromOperation`, () => {
  const firstId = `gid://shopify/Collection/12345`
  const secondId = `gid://shopify/Collection/54321`
  const firstProductId = `gid://shopify/Product/22345`
  const secondProductId = `gid://shopify/Product/32345`
  const thirdProductId = `gid://shopify/Product/64321`

  const bulkResults = [
    {
      id: firstId,
    },
    {
      id: firstProductId,
      __parentId: firstId,
    },
    {
      id: secondId,
    },
    {
      id: secondProductId,
      __parentId: firstId,
    },
    {
      id: thirdProductId,
      __parentId: secondId,
    },
  ]

  afterEach(() => {
    server.resetHandlers()
  })

  beforeEach(() => {
    server.use(
      graphql.query<ICurrentBulkOperationResponse>(
        `OPERATION_STATUS`,
        resolveOnce({
          currentBulkOperation: {
            id: `12345`,
            status: `COMPLETED`,
          },
        })
      ),
      graphql.mutation<IBulkOperationRunQueryResponse>(
        `INITIATE_BULK_OPERATION`,
        resolveOnce({
          bulkOperationRunQuery: {
            bulkOperation: {
              id: `12345`,
              objectCount: `0`,
              query: ``,
              status: `CREATED`,
              url: ``,
            },
            userErrors: [],
          },
        })
      ),
      graphql.query<{ node: IBulkOperationNode }>(
        `OPERATION_BY_ID`,
        resolveOnce({
          node: {
            status: `CREATED`,
            id: ``,
            objectCount: `0`,
            query: ``,
            url: ``,
          },
        })
      ),
      graphql.query<{ node: IBulkOperationNode }>(
        `OPERATION_BY_ID`,
        resolve({
          node: {
            status: `COMPLETED`,
            id: `12345`,
            objectCount: `1`,
            query: ``,
            url: BULK_DATA_URL,
          },
        })
      )
    )
  })

  const config = [
    {
      bulkResults: [
        {
          id: `gid://shopify/Product/6950279086286`,
        },
        {
          id: `gid://shopify/MediaImage/22755647652046`,
        },
        {
          id: `gid://shopify/Metafield/21077562228942`,
          __parentId: `gid://shopify/Product/6950279086286`,
        },
        {
          id: `gid://shopify/Product/6966721052878`,
        },
        {
          id: `gid://shopify/Metafield/21078060335310`,
          __parentId: `gid://shopify/Product/6966721052878`,
        },
      ],
      prioritize: true,
      type: `products`,
    },
  ]

  config.forEach(({ bulkResults, prioritize, type }) => {
    it(`testing prioritize: ${prioritize}, type: ${type}`, async () => {
      server.use(
        rest.get(BULK_DATA_URL, (req, res, ctx) => res(ctx.json(bulkResults)))
      )
      const options = {
        password: ``,
        storeUrl: `my-shop.shopify.com`,
        downloadImages: true,
      }
      const operations = createOperations(mockGatsbyApi(), options)

      const finishLastOperation = jest.spyOn(operations, `finishLastOperation`)
      const completedOperation = jest.spyOn(operations, `completedOperation`)
      const cancelOperationInProgress = jest.spyOn(
        operations,
        `cancelOperationInProgress`
      )
      const gatsbyApi = mockGatsbyApi()
      const pluginOptions = mockPluginOptions()

      const sourceFromOperation = makeSourceFromOperation(
        finishLastOperation,
        completedOperation,
        cancelOperationInProgress,
        gatsbyApi,
        { ...pluginOptions, prioritize }
      )

      await sourceFromOperation(operations.productsOperation)

      const { createNode } = gatsbyApi.actions

      expect(createNode).toHaveBeenCalledTimes(bulkResults.length)
      expect(cancelOperationInProgress.mock.calls.length).toEqual(
        prioritize ? 1 : 0
      )
      expect(finishLastOperation.mock.calls.length).toEqual(prioritize ? 0 : 1)
    })
  })

  for (const prioritize of [true, false]) {
    for (const type of nodeTypes) {
      it(generateTestName(prioritize, type), async () => {
        const fileContents = fs.readFileSync(
          path.join(__dirname, `fixtures`, `bulk-results`, `${type}.jsonl`),
          { encoding: `utf8` }
        )

        const lineCountInFile = countJsonLRows(fileContents)

        server.use(
          rest.get(BULK_DATA_URL, (req, res, ctx) =>
            res(ctx.text(fileContents))
          )
        )

        const options = {
          password: ``,
          storeUrl: `my-shop.shopify.com`,
          downloadImages: true,
        }
        const operations = createOperations(mockGatsbyApi(), options)

        const finishLastOperation = jest.spyOn(
          operations,
          `finishLastOperation`
        )
        const completedOperation = jest.spyOn(operations, `completedOperation`)
        const cancelOperationInProgress = jest.spyOn(
          operations,
          `cancelOperationInProgress`
        )
        const gatsbyApi = mockGatsbyApi()
        const pluginOptions = mockPluginOptions()

        const sourceFromOperation = makeSourceFromOperation(
          finishLastOperation,
          completedOperation,
          cancelOperationInProgress,
          gatsbyApi,
          { ...pluginOptions, prioritize }
        )

        await sourceFromOperation(operations.productsOperation)

        const { createNode } = gatsbyApi.actions

        expect(createNode).toHaveBeenCalledTimes(lineCountInFile)
        expect(cancelOperationInProgress.mock.calls.length).toEqual(
          prioritize ? 1 : 0
        )
        expect(finishLastOperation.mock.calls.length).toEqual(
          prioritize ? 0 : 1
        )
        // expect(createNode).toHaveBeenCalledWith(
        //   expect.objectContaining({
        //     id: `gid://shopify/Product/6950279086286`,
        //   })
        // )
      })
    }
  }
})
