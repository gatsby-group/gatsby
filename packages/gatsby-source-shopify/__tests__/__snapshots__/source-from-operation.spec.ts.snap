// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeSourceFromOperation returns correctly when running a non-priority build 1`] = `Array []`;
